java/ql/src/semmle/code/java/security/RequestForgery.qll

```diff
@@ -0,0 +1,334 @@
+ * An argument to an `org.springframework.http.RequestEntity` constructor call
+ *   which is an URI taken as a sink for request forgery vulnerabilities.
+ */
+private class RequestEntityUriArg extends RequestForgerySink {
+  RequestEntityUriArg() {
+    exists(ClassInstanceExpr e, Argument a |
+      e.getConstructedType() instanceof SpringRequestEntity and
+      e.getAnArgument() = a and
+      a.getType() instanceof TypeUri and
+      this.asExpr() = a
+    )
+  }
+}
+
+/**
+ * A Spring Rest Template method
+ * which take a URL as an argument.
+ */
+private class SpringRestTemplateUrlMethod extends Method {
+  SpringRestTemplateUrlMethod() {
+    this.getDeclaringType() instanceof SpringRestTemplate and
+    this.hasName([
+        "doExecute", "postForEntity", "postForLocation", "postForObject", "put", "exchange",
+        "execute", "getForEntity", "getForObject", "patchForObject"
+      ])
+  }
+
+  /**
+   * Gets the argument which corresponds to a URL argument
+   * passed as a `java.net.URL` object or as a string or the like
+   */
+  Argument getUrlArgument(MethodAccess ma) {
+    // doExecute(URI url, HttpMethod method, RequestCallback requestCallback,
+    //  ResponseExtractor<T> responseExtractor)
+    result = ma.getArgument(0)
+  }
+}
+
+/** A sanitizer for request forgery vulnerabilities. */
+abstract class RequestForgerySanitizer extends DataFlow::Node { }
+
+private class PrimitiveSanitizer extends RequestForgerySanitizer {
+  PrimitiveSanitizer() { this.getType() instanceof PrimitiveType }
+}
+
+private class HostnameSanitizingPrefix extends CompileTimeConstantExpr {
+  int offset;
+
+  HostnameSanitizingPrefix() {
+    // Matches strings that look like when prepended to untrusted input, they will restrict
+    // the host or entity addressed: for example, anything containing `?` or `#`, or a slash that
+    // doesn't appear to be a protocol specifier (e.g. `http://` is not sanitizing), or specifically
+    // the string "/".
+    exists(
+      this.getStringValue()
+          .regexpFind(".*([?#]|[^?#:/\\\\][/\\\\]).*|[/\\\\][^/\\\\].*|^/$", 0, offset)
+    )
+  }
+
+  /**
+   * Gets the offset in this constant string where a sanitizing substring begins.
+   */
+  int getOffset() { result = offset }
+}
+
+private AddExpr getParentAdd(AddExpr e) { result = e.getParent() }
+
+private AddExpr getAnAddContainingHostnameSanitizingPrefix() {
+  result = getParentAdd*(any(HostnameSanitizingPrefix p).getParent())
+}
+
+private Expr getASanitizedAddOperand() {
+  exists(AddExpr e |
+    e = getAnAddContainingHostnameSanitizingPrefix() and
+    (
+      e.getLeftOperand() = getAnAddContainingHostnameSanitizingPrefix() or
+      e.getLeftOperand() instanceof HostnameSanitizingPrefix
+    ) and
+    result = e.getRightOperand()
+  )
+}
+
+private MethodAccess getNextAppend(MethodAccess append) {
+  result = any(StringBuilderVar sbv).getNextAppend(append)
+}
+
+private Expr getQualifier(MethodAccess e) { result = e.getQualifier() }
```

Comment:
This predicate seems very superfluous.

Reply:
Also unused. Probably a part-done cleanup.

Reply:
This is used in `getQualifier*(initializer)`, where this phrasing allows me to iterate downcasting to MethodAccess and then calling getQualifier. Is there a better way?

Reply:
How about `methodAccess.getQualifier*()`? The kleene-star HOP is also applicable to member predicates.

Reply:
Hmm, I'm actually unsure whether that will downcast from `Call` to `MethodAccess`, but I don't know whether that's important here?

Reply:
Actually, this might need to be
```suggestion
private Expr getQualifier(Expr e) { result = e.(MethodAccess).getQualifier() }
```
to support the reflective case below without restricting `initializer = constructor` to be a `MethodAccess`.